pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ReferenceResolver is an interface for resolving cross resource attribute references.
+// It should be implemented by strcuts that are used for custom resources feilds.
+// See #874 for more information
+type ReferenceResolver interface {
+	// Resolve retrieve and builds the referenced attribute
+	Resolve(reader client.Reader) (value interface{}, err error)
+
+	// Assign accepts a resource object, and assigns the given value to the corresponding property
+	Assign(resource interface{}, value interface{}) error
+}
+
+// ResolveReferencers resolves references made to other resources
+func ResolveReferencers(ctx context.Context, mgd Managed, c client.Reader) (bool, error) {
+	resolvers, err := findRefResolverFields(mgd, false)
+	if err != nil {
+		return false, err
+	}
+
+	allRferencesReady, err := checkIfAllReferencesAreReady(getAllObjectReferences(resolvers))
+	if err != nil || !allRferencesReady {
+		return false, err
+	}
+
+	for _, r := range resolvers {
+
+		val, err := r.Resolve(c)
+		if err != nil {
+			return allRferencesReady, err
+		}
+
+		err = r.Assign(mgd, val)
+		if err != nil {
+			return allRferencesReady, err
+		}
+	}
+
+	return false, nil
+}
+
+// findRefResolverFields recursively finds all non-nil fields in a struct and its sub-structs
+// that implement ReferenceResolver and have `managed_resource:"reference_resolver"` tag
+func findRefResolverFields(obj interface{}, objHasTheRightTag bool) ([]ReferenceResolver, error) { // nolint:gocyclo
+
+	v := reflect.ValueOf(obj)
+	result := []ReferenceResolver{}
+
+	// ignore nil fields
+	if v.Kind() == reflect.Ptr && v.IsNil() {
+		return nil, nil
+	}
+
+	if objHasTheRightTag {
+		if ar, implements := obj.(ReferenceResolver); implements {
+			return []ReferenceResolver{ar}, nil
+		} else if v.Kind() == reflect.Struct {
+			return nil, fmt.Errorf("the field has the 'reference_resolver' tag, but has not implemented ReferenceResolver interface")
+		}
+	}
+
+	switch v.Kind() {
+	case reflect.Struct:
+		for i := 0; i < v.NumField() && v.Field(i).CanInterface(); i++ {
+			val, ok := reflect.TypeOf(obj).Field(i).Tag.Lookup("managed_resource")
+			hasTheRightTag := ok && val == "reference_resolver"
```

Comment:
These strings technically are called [`packages`](https://github.com/golang/go/commit/25733a94fde4f4af4b6ac5ba01b7212a3ef0f013). 
And since there related to `resource` package, I gave them the more specific name `managed_resource`. I prefer to keep it this way as `crossplane` is less informative when the code itself is in Crossplane project.